contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol

```diff
@@ -0,0 +1,195 @@
+import "@openzeppelin-4.5/contracts/utils/cryptography/ECDSA.sol";
+import "./ISignVerifierRegistry.sol";
+
+pragma solidity 0.8.9;
+
+/**
+ * @title Powered by Art Blocks ERC-721 core contract.
+ * Support for IYK pull mechanism to transfer tokens based on verified signatures
+ * @notice Due to the pull mechanism (claimNFT) being the only supported transfer function, these NFTs will NOT be tradeable on secondary marketplaces
+ * @author Art Blocks Inc.
+ */
+contract GenArt721CoreV2_IYK is GenArt721CoreV2_PBAB {
+    using ECDSA for bytes32;
+
+    // Address of registry that resolves the signVerifier
+    ISignVerifierRegistry public signVerifierRegistry;
+    bytes32 public signVerifierId;
+
+    // Nonce per user, used to prevent replay signature attacks
+    mapping(address => uint256) public claimNonces;
+
+    /**
+     * @notice Initializes contract.
+     * @param _tokenName Name of token.
+     * @param _tokenSymbol Token symbol.
+     * @param _randomizerContract Randomizer contract.
+     * @param _startingProjectId The initial next project ID.
+     * @param _signVerifierRegistry Address of registry that resolves the signVerifierId.
+     * @param _signVerifierId ID of the signVerifier.
+     * @dev _startingProjectId should be set to a value much, much less than
+     * max(uint256) to avoid overflow when adding to it.
+     */
+    constructor(
+        string memory _tokenName,
+        string memory _tokenSymbol,
+        address _randomizerContract,
+        uint256 _startingProjectId,
+        address _signVerifierRegistry,
+        bytes32 _signVerifierId
+    )
+        GenArt721CoreV2_PBAB(
+            _tokenName,
+            _tokenSymbol,
+            _randomizerContract,
+            _startingProjectId
+        )
+    {
+        signVerifierRegistry = ISignVerifierRegistry(_signVerifierRegistry);
+        signVerifierId = _signVerifierId;
+    }
+
+    /**
+     * @notice Transfers a token from its owner to the recipient
+     * @param _sig The result of getClaimSigningHash signed by the signVerifier
+     * @param _blockExpiry The block as of which the signature is no longer valid
+     * @param _recipient The address that receives the token
+     * @param _tokenId The tokenId being claimed
+     * @dev ECDSA signatures are used to verify the permission to claim a NFT
+     */
+    function claimNFT(
+        bytes memory _sig,
+        uint256 _blockExpiry,
+        address _recipient,
+        uint256 _tokenId
+    ) public virtual {
+        bytes32 message = getClaimSigningHash(
+            _blockExpiry,
+            _recipient,
+            _tokenId
+        ).toEthSignedMessageHash();
+        require(
+            ECDSA.recover(message, _sig) == getSignVerifier(),
+            "Permission to call this function failed"
+        );
+        require(block.number < _blockExpiry, "Sig expired");
+
+        address from = ownerOf(_tokenId);
+        require(from != address(0));
+
+        claimNonces[_recipient]++;
+
```

Comment:
Since `claimNonces` gets updated to prevent reusing the current signature, is this necessary? Would prefer to not add the gas cost here unless there's a strong reason to.

Reply:
Yeah, I agree that this idea doesn't do anything worthwhile, and just adds gas. Resolving!